### DISCUSSION TITLE
MooseDocs w/ Animal Codes

### DISCUSSION THREAD

**Discussion by roskofnj** [idaholab/moose#17436] [Q&A Tools] [answered]
Each Animal code has it's own moosedoc.py which is used to start a document server on the localhost (127.0.0.1:8000). I would like to have all the Animal code document servers running at the same time. However I cannot launch multiple document servers on different ports - when the second doc server is launched it simply overwrites the first.

Is there a way to start a document server that contains all documentation for all Animal codes, as well as RELAP7? It seems strange that I cannot start different servers on different ports. 

Thanks very much for any help you're able to offer.

**Comment by cticenhour** [comment]
You can change the port for the MooseDocs build command using the `--port` option. So, if I wanted to change my documentation website port to `9000`, I would just run the following in my app's `doc` directory:

```
./moosedocs.py build --port 9000 --serve
```

and the built pages would be available on `http://127.0.0.1:9000`. To see all of the options for the build command, type `./moosedocs.py build -h` in your `doc` directory.

**Reply by roskofnj** [reply]
This is exactly what I've trying to do. But any time I try to launch a separate instance it overwrites the first. For example:
1) in the Griffin directory, I'll do `./moosedocs.py build --serve --port 8001` and then Griffin document server launches correctly at 127.0.0.1:8001
2) in the Sockeye directory I'll do `./moosedocs.py build --serve --port 8002` and the Sockeye document server correctly launches at 127.0.0.1:8002. But then if I visit 127.0.0.1:8001 (which used to be a Griffin doc server) it now shows Sockeye.

So it seems like when I launch a second server it is always overwriting the first.

**Reply by cticenhour** [reply]
I deleted my previous comment, as I just got to test the `--host` alternative option (setting each documentation instance as a separate local IP) on a MacOS machine. Because both `127.x.x.x` URLs are loopback interfaces (referring to the same machine), it will be overwritten regardless of the IP difference. @loganharbour's option is the proper answer here.

**Reply by roskofnj** [reply]
Thanks very much. I actually tried just what you were originally thinking. 

I way out of my league on this server stuff, so I really appreciate you both (@loganharbour ) for bearing with me on this.

**Reply by loganharbour** [reply]
Thanks for coming to us with questions! This information may be able to help someone in the future.

I'd also like to add that in terms of documentation, if you have any suggestions to make the process any clearer to the user, we welcome issues on GitHub!

**Comment by loganharbour** [comment, marked as answer]
Note that while the serve option is the easiest way to get the documentation going, it isn't required to view the documentation. We suggest using `--serve` because (1) it is convenient, and (2) it auto updates from source so that you can make changes and see the changes almost real time.

If you would like to just generate the raw static site in one go (which you can also open with your browser by opening the html), you can generate with

```
./moosedocs.py build --destination <directory>
```

in which you fill `<directory>` with a directory where you want the site to be built into. Once built, you can traverse into said directory, start with the `index.html` page and browse as needed without having to start the local service to host the site. You could do this for all of the applications that you need documentation for, and put them in a convenient location for easy access.

**Reply by roskofnj** [reply]
Great, I appreciate this alternative approach. I'll give it a try. might be easier and less prone to issues than my original attempt. At this point I don't think I'll be doing too many changes to the documentation, so it should be sufficient.
